tools/pre-commit-githook

```diff
@@ -73,13 +73,122 @@ not_pruned () {
+
+AUTOCONF_FILES=\
+'configure configure.ac VERSION aclocal.m4 build-aux/* '\
+'tools/autogen tools/git-dev-options.sh'
+
+# Convert $AUTOCONF_FILES to a BRE
+PATHS="${AUTOCONF_FILES//./\\.}"
+PATHS="${PATHS//\*/.*}"
+PATHS="${PATHS// /\\|}"
+
+OVERRIDE_MESSAGE='(you can override githooks with git-commit --no-verify)'
+WRONG_AUTOCONF=0
+
+if git diff --diff-filter=d --staged --name-only | grep -q "'$PATHS'" ; then
+  # Get the AC_PREREQ line in configure.ac for the required autoconf version
+  PREREQ="$(git cat-file --textconv :configure.ac \
+              | sed -ne 's/^AC_PREREQ(\[\(.*\)\])/\1/p')"
+  if [[ -z $PREREQ ]]; then
+    echo 'Unable to find/parse the AC_PREREQ macro in configure.ac'
+    echo 'This line should be of the form AC_PREREQ([2.69])'
+    echo '(with no whitespace or comment)'
+    STATUS=1
+  else
+    # Check for autoconf and its version
+    AUTOCONF_TOOL='autoconf'
+    # Check version of autoconf
+    set -o pipefail
+    AUTOCONF_VERSION="$($AUTOCONF_TOOL --version 2>/dev/null | head -n 1)"
+    if [[ $? -ne 0 ]]; then
+      echo 'Files affecting configure updated, but autoconf not found'
+      echo 'Unable to verify that configure is up-to-date'
+      echo "$OVERRIDE_MESSAGE"
+      STATUS=1
+    else
+      AUTOCONF_VERSION="${AUTOCONF_VERSION##* }"
+      if [[ $AUTOCONF_VERSION != $PREREQ ]]; then
+        # Found autoconf, but it's the wrong version. If it's older,
+        # tools/autogen will fail. If it's newer, it may succeed, but CI may
+        # fail. autoconf is frequently available at an exact version, so try
+        # the two known names for it.
+        for tool in $AUTOCONF_TOOL-$PREREQ $AUTOCONF_TOOL$PREREQ; do
+          VERSION="$($tool --version 2>/dev/null | head -n 1)"
+          if [[ $? -eq 0 ]]; then
+            VERSION="${VERSION##* }"
+            if [[ $VERSION != $PREREQ ]]; then
+              continue
+            fi
+          else
+            continue
+          fi
+          echo "autoconf has version $AUTOCONF_VERSION; using $tool instead"
+          AUTOCONF_TOOL="$tool"
+          AUTOCONF_VERSION="$VERSION"
+          break
+        done
+      fi
+
+      if [[ $AUTOCONF_VERSION != $PREREQ ]]; then
+        # We're using the wrong version of autoconf: if all other tests succeed,
+        # display a warning that CI may complain (CI uses the version specified
+        # by AC_PREREQ).
+        WRONG_AUTOCONF=1
+      fi
+
+      # Checkout the relevant files from the index to a temporary directory to
+      # test them.
+      BUILD_DIR="$(mktemp -d)"
+      BUILD_DIR="${BUILD_DIR%%/}/"
+      if [[ -z $BUILD_DIR ]]; then
+        echo 'Unable to create a temporary directory to test configure.ac'
+        STATUS=1
+      else
+        git checkout-index --prefix "$BUILD_DIR" -- $AUTOCONF_FILES
+        pushd "$BUILD_DIR" > /dev/null
+        mkdir -p a b
+        mv configure a/
+        echo 'Regenerating configure...'
+        if tools/autogen "$AUTOCONF_TOOL"; then
```

Comment:
Could we pass our `AUTOCONF_TOOL` preference through the environment, to make both the caller and the callee more readable?

Reply:
We could, but to be honest (biased from a Windows perspective, possibly) I find using environment variables like this to be truly dreadful (and have lost countless hours debugging elsewhere in opam where this was thought to be a good idea...). It's effectively choosing global variables instead of function parameters. Whatever name we pick, there's then a global variable which can cause `tools/autogen` to fail?